File: kotlin-electron/karakum/plugins/convertSkippedGenerics.js
import ts from "typescript";

export default function (node, context, render) {
    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "Function"
        && !node.typeArguments
    ) {
        if (
            ts.isParameter(node.parent)
            && ts.isIdentifier(node.parent.name)
            && node.parent.name.text === "listener"
        ) {
            return `${render(node.typeName)}<Unit>`
        }

        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "Event"
        && !node.typeArguments
    ) {
        if (ts.isTypeAliasDeclaration(node.parent)) {
            return `${render(node.typeName)}`
        }

        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text === "Uint8Array"
        && !node.typeArguments
    ) {
        return `${render(node.typeName)}<*>`
    }

    if (
        ts.isExpressionWithTypeArguments(node)
        && ts.isIdentifier(node.expression)
        && node.expression.text === "Event"
        && !node.typeArguments
    ) {
        return `${render(node.expression)}<Any>`
    }

    return null
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/plugins/convertSkippedGenerics.kt
package electron.karakum.plugins

import electron.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.isExpressionWithTypeArguments
import typescript.isIdentifier
import typescript.isParameter
import typescript.isTypeAliasDeclaration
import typescript.isTypeReferenceNode

val convertSkippedGenerics = createPlugin { node, _, render ->
    nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Function")
        ensure(node.typeArguments == null)

        nullable {
            val parameter = ensureNotNull(node.getParentOrNull())
            ensure(isParameter(parameter))

            val parameterName = ensureNotNull(parameter.name)
            ensure(isIdentifier(parameterName))
            ensure(parameterName.text == "listener")

            "${render(node.typeName)}<Unit>"
        } ?: run {
            "${render(node.typeName)}<*>"
        }
    } ?: nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Event")
        ensure(node.typeArguments == null)

        nullable {
            val typeAlias = ensureNotNull(node.getParentOrNull())
            ensure(isTypeAliasDeclaration(typeAlias))

            render(node.typeName)
        } ?: run {
            "${render(node.typeName)}<*>"
        }
    } ?: nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Uint8Array")
        ensure(node.typeArguments == null)

        "${render(node.typeName)}<*>"
    } ?: nullable {
        ensure(isTypeReferenceNode(node))

        val typeName = node.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Buffer")
        ensure(node.typeArguments == null)

        "${render(node.typeName)}<*>"
    } ?: nullable {
        ensure(isExpressionWithTypeArguments(node))

        val expression = node.expression
        ensure(isIdentifier(expression))
        ensure(expression.text == "Event")
        ensure(node.typeArguments == null)

        "${render(node.expression)}<Any>"
    }
}
